### DISCUSSION TITLE
External Proxy not coming across in payload

### DISCUSSION THREAD

**Discussion by phil-hotchkiss** [mia-platform/community#393] [Q&A] [answered]
### What product are you having troubles with?

Console

### What Console version are you using?

12.2.0

### Description

My project is using the Enhanced Workflow with the pull based Deployment Strategy. The payload that comes across is not including the defined external proxies. 

When I look at an example of the api-console-config.json that was produced during our proof of concept and used the push based (mlp) deployment model, I can see that the external proxies came across in that api-console-config.json as one of the defined services. 

Was this missed during the Enhanced Workflow implementation as a service that needs to be included in the payload? 

For context, here is what I'm comparing. In payload.json, I have a defined proxy named tapi that is missing. In the api-console-config.json, you can see I have a service of type external that is pointing at a proxy. 

payload.json: 
![image](https://github.com/mia-platform/community/assets/136851567/155808a9-4558-4dbe-b92f-9b2c0d05b20a)

api-console-config.json
![image](https://github.com/mia-platform/community/assets/136851567/d23b21ae-db42-44c3-8c1a-98dbafedd671)

Please advise on how to proceed. 

Thanks!

### Actual Outcome

The payload.json does not include the http proxy definitions that are defined in the console in the enhanced workflow (pull-based deployment strategy)

### Expected Outcome

I expect the external proxies to be part of the payload that is provided as a webhook in enhanced workflow.

**Comment by fredmaggiowski** [comment, marked as answer]
Hi @phil-hotchkiss, external proxies are a feature that's implemented by the api gateway (both nginx and envoy based gateways); therefore, even if they appear to be within the services object when using the standard workflow (the one you used during the poc) they are not part of the services definition for the external generator payload but are already set up in the api gateway configuration files

**Reply by phil-hotchkiss** [reply]
@fredmaggiowski Does this infer that I need to map `Endpoints` in order for the external proxies to show up in the envoy configuration? Or am I supposed to be getting those by having an external proxy defined? 

We understood the endpoints to be reserved for behaviors of the project you intend to expose to other projects or clients. We are intending to leverage this external proxy to define egress behavior that one of our Microservices needs. 

Let me know how to proceed! Thanks!

**Reply by fredmaggiowski** [reply]
Yes, you need to map necessary endpoints to be able to leverage proxies; using them as egress is definitely feasible, I'd suggest making sure you are correctly protecting the endpoints from external accesses. Another way to centralize egress could be using the [http proxy manager](https://docs.mia-platform.eu/docs/runtime_suite/http-proxy-manager/overview) service?

**Reply by phil-hotchkiss** [reply]
So, once I've defined an endpoint as such: 
![image](https://github.com/mia-platform/community/assets/136851567/10bfb952-bfa4-4bef-9308-ac7812f4ebed)

How should we call this endpoint within a microservice? Our initial expectation was that it would be a named service that we could call, but because it's defined as an endpoint on the project, I'm not following how the microservices within the project should reference the above configuration. 

Thanks!

**Reply by fredmaggiowski** [reply]
The service exposing it is the api-gateway itself. Assuming you are using the envoy-based gateway (you have listeners) with a service named `api-gateway` you should be able to use the proxy by calling `http(s)://api-gateway/tasks`. Does it work?

**Reply by phil-hotchkiss** [reply]
That makes total sense. I appreciate the help.
